sea: snapshot support in single executable applications

joyeecheung opened the pull request:
Refs: https://github.com/nodejs/single-executable/discussions/57

This patch adds snapshot support to single executable applications.
To build a snapshot from the main script when preparing the
blob that will be injected into the single executable application,
add `"useSnapshot": true` to the configuration passed to
`--experimental-sea-config`. For example:

```
{
    "main": "snapshot.js",
    "output": "sea-prep.blob",
    "useSnapshot": true
}
```

The main script used to build the snapshot must invoke
`v8.startupSnapshot.setDeserializeMainFunction()` to configure the
entry point. The generated startup snapshot would be part of the
preparation blob and get injected into the final executable.

When the single executable application is launched, instead of running
the `main` script from scratch, Node.js would instead deserialize the
snapshot to get to the state initialized during build-time directly.

```
$ /path/to/node --experimental-sea-config sea-config.json
$ cp /path/to/node sea
$ postject sea NODE_SEA_BLOB sea-prep.blob --sentinel-fuse NODE_SEA_FUSE_fce680ab2cc467b6e072b8b5df1996b2
$ ./sea
```

<!--
Before submitting a pull request, please read
https://github.com/nodejs/node/blob/HEAD/CONTRIBUTING.md.

Commit message formatting guidelines:
https://github.com/nodejs/node/blob/HEAD/doc/contributing/pull-requests.md#commit-message-guidelines

For code changes:
1. Include tests for any bug fixes or new features.
2. Update documentation if relevant.
3. Ensure that `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

RaisinTen commented:
Thanks for doing this!

> add some tooling to process the inputs

I personally really like this approach. I've opened a discussion in https://github.com/nodejs/single-executable/discussions/58, so that we can have a decision on this.

joyeecheung commented:
Rebased using the new SnapshotBuilder::Generate() overload added in https://github.com/nodejs/node/pull/48242 and added a test. Will add more doc and tests next week but the implementation is there.

piranna commented:
> Rebased using the new SnapshotBuilder::Generate() overload added in #48242 and added a test. Will add more doc and tests next week but the implementation is there.

What's the difference between snapshot and code cache (bytenode)?

joyeecheung commented:
> What's the difference between snapshot and code cache (bytenode)?

If you have a bunch of top level code with a bunch of inner functions, usually you run all the top level code and compile some of the inner functions, the snapshot contains the state of the heap after the top-level code is run and the bytecode of the inner functions that would be run again later. When deserializing the snapshot, you deserialize the state of an initialized heap (so there is no need to run the top-level code at all), and maybe later when the application invokes some of the inner functions V8 deserialize the bytecode in the snapshot during the compilation of the function to speed things up. If there's just the code cache, what usually happens is that you have the bytecode of the top-level code and some inner functions. At runtime V8 deserializes the bytecode in the code cache to speed up compilation of the top-level code, but it still needs to run the top-level code to set up the heap.

piranna commented:
> If you have a bunch of top level code with a bunch of inner functions, usually when building the snapshot you run all the top level code and compile some of the inner functions, the snapshot contains the state of the heap after the top-level code is run and the bytecode of the inner functions that would be run again later. When deserializing the snapshot, you deserialize the state of an initialized heap (so there is no need to run the top-level code at all), and maybe later when the application invokes some of the inner functions, V8 deserialize the bytecode in the snapshot during the compilation of the function to speed things up. If there's just the code cache, what usually happens is that you have the bytecode of the top-level code and some inner functions. At runtime V8 deserializes the bytecode in the code cache to speed up compilation of the top-level code, but it still needs to run the top-level code to set up the heap.

Thank you @joyeecheung for your detailed explain :-) So, if I understood it correctly, a snapshoot is a combo of a module compiled and bytecode functions plus heap after executing the module top level code, while the code cache only have the bytecode of all the module code, isn't it?

joyeecheung commented:
> So, if I understood it correctly, a snapshoot is a combo of a module compiled and bytecode functions plus heap after executing the module top level code, while the code cache only have the bytecode of all the module code, isn't it?

Roughly yes. V8 does not always emit bytecode for all the code it encounters (e.g. inner-inner functions are usually just parsed for syntax checks, but no bytecode is generated until they actually get executed). In some cases snapshots may contain more bytecode if during the snapshot building cases these functions are invoked.

piranna commented:
> V8 does not always emit bytecode for all the code it encounters

Maybe [bytenode](https://github.com/bytenode/bytenode) is forcing its generation? Could it of our interest when generating the snapshots, so no Javascript code is included in it?

benjamingr commented:
@RaisenTen can you please confirm you are comfortable with this based on https://github.com/nodejs/single-executable/discussions/58 and https://github.com/nodejs/single-executable/discussions/58#discussioncomment-5565539 landing

joyeecheung commented:
>  Could it of our interest when generating the snapshots, so no Javascript code is included in it?

I would say that's not always a win, you could also have a lot of dead code in that application, then eagerly generating code may result in a memory/size overhead that's not always worth it. As a configurable thing that might be useful though.

RaisinTen commented:
I can try to take a look day after tomorrow. (I'll be traveling pretty much the entire day tomorrow)

piranna commented:
> As a configurable thing that might be useful though.

Agree on that 👍🏻

RaisinTen commented on src/node_sea.h:
```diff
@@ -21,19 +23,24 @@ const uint32_t kMagic = 0x143da20;
 enum class SeaFlags : uint32_t {
   kDefault = 0,
   kDisableExperimentalSeaWarning = 1 << 0,
+  kuseSnapshot = 1 << 1,
```
```suggestion
  kUseSnapshot = 1 << 1,
```

RaisinTen commented on src/node_snapshotable.cc:
```diff
@@ -584,7 +584,7 @@ size_t SnapshotSerializer::Write(const SnapshotMetadata& data) {
 // [    ...       ]  code_cache
 
-std::vector<char> SnapshotData::ToBlob() const {
+void SnapshotData::ToBlob(std::vector<char>* out) const {
```
Why do we need this change? In `src/node_sea.cc`, we can do `snapshot_blob = snapshot.ToBlob();` and keep using the existing API instead of creating another one, right?

joyeecheung replied:
I did it to avoid the copy (which would be there if we just do the assign), though I think we could probably do a move assignment...

RaisinTen commented on doc/api/single-executable-applications.md:
```diff
@@ -177,14 +181,33 @@ The configuration currently reads the following top-level fields:
+
+When `useSnapshot` is set to true in the configuration, during the generation
+of the single executable preparation blob, Node.js will run the `main` script
+to generate a startup snapshot. The script must invoke
+[`v8.startupSnapshot.setDeserializeMainFunction()`][] to set up the entry point.
+The generated startup snapshot would be part of the preparation blob and get
+injected into the final executable. When the single executable application is
```
Would it be possible to generate and use the snapshot in a way that doesn't require users to use the `v8.startupSnapshot.setDeserializeMainFunction()` function in their code? I think that would help in making this easier to use because that way users wouldn't be required to change the SEA code to use this feature, they just need to set `useSnapshot` to `true` and that should be it AFAICT.

joyeecheung replied:
Using snapshot should involve intentional effort to split the application code into initialization code that can be run during snapshot-building time and application code that can be run during run-time. When `useSnapshot` is true, the script is run *at build time* to produce the snapshot, and it's expected to run to completion. So if user does not configure additional code to be run at run-time...then we don't have anything to run when the SEA is launched.

If what we want is not having to do code changes, one way to do it is to add another `snapshot_script` field so that a separate snapshot script is run at build time and the main script is left to be run at run-time.  But I don't think that's necessarily better than requiring `v8.startupSnapshot.setDeserializeMainFunction()`, because then it's harder for the main script to retrieve states set up by the snapshot script without, um, using any JS API explicitly (unless the two scripts use some global objects to communicate, which is not ideal). And when snapshot is used it's usually useful or necessary to be aware about it and fix up the states, or do things differently, it's not designed to be something that can just be configured with a boolean and be it IMO.

joyeecheung replied:
ping @RaisinTen does that answer your question?

RaisinTen commented on src/node_sea.cc:
```diff
@@ -235,10 +247,23 @@ std::optional<SeaConfig> ParseSingleExecutableConfig(
-ExitCode GenerateSingleExecutableBlob(const SeaConfig& config) {
+ExitCode GenerateSingleExecutableBlob(
+    const SeaConfig& config,
+    const std::vector<std::string> args,
+    const std::vector<std::string> exec_args) {
```
```suggestion
    const std::vector<std::string>& args,
    const std::vector<std::string>& exec_args) {
```
Can these be const refs?

RaisinTen commented on src/node_sea.cc:
```diff
@@ -248,7 +273,42 @@ ExitCode GenerateSingleExecutableBlob(const SeaConfig& config) {
+      static_cast<bool>(config.flags & SeaFlags::kUseSnapshot);
+  if (builds_snapshot_from_main) {
+    SnapshotData snapshot;
+    std::vector<std::string> patched_args = {args[0], GetAnonymousMainPath()};
```
Should we use the executable filename / the original script name (the code_path variable being added in https://github.com/nodejs/node/pull/48191) instead for the main script file name?

RaisinTen commented on src/node_sea.cc:
```diff
@@ -248,7 +273,42 @@ ExitCode GenerateSingleExecutableBlob(const SeaConfig& config) {
   }
 
-  SeaResource sea{config.flags, main_script};
+  std::vector<char> snapshot_blob;
```
What do you think about moving the snapshot generation part into a separate function for more modularity?

RaisinTen commented on src/node_sea.cc:
```diff
@@ -235,10 +247,56 @@ std::optional<SeaConfig> ParseSingleExecutableConfig(
+                                const std::string& main_script,
+                                std::vector<char>* snapshot_blob) {
+  SnapshotData snapshot;
+  std::vector<std::string> patched_args = {args[0], config.main_path};
```
Should we also append the list of CLI args that comes after `node --experimental-sea-config sea-config.json`, i.e., all elements of `args` after index `0`, to this vector? I'm saying that because when we run `node --snapshot-blob snapshot.blob --build-snapshot snapshot.js a b c`, it persists `a`, `b` and `c` in the list of CLI args present during the snapshot building phase.
```suggestion
  std::vector<std::string> patched_args = args;
  patched_args.insert(patched_args.begin() + 1, config.main_path);
```

joyeecheung replied:
> it persists a, b and c in the list of CLI args present during the snapshot building phase.

It probably does not matter because we will always refresh the arguments after the snapshot is deserialized. But I think it might be useful to pass the arguments for testing purposes (usually users should not configure any of them during snapshot building, we generally refuse to load the snapshot if the flags do not match, except for certain known safe flags).

RaisinTen replied:
> It probably does not matter because we will always refresh the arguments after the snapshot is deserialized.

Okay, I had assumed that it was expected behavior because earlier I had come across a blog post (unofficial) that tried to store some of the CLI args - https://blog.logrocket.com/snapshot-flags-node-js-v18-8/

joyeecheung replied:
Thinking about it again this seems to be a bit weird because the main script is configured in the config, while the command line arguments are passed from the command line. I think we should probably add a programmatic API or a JSON config field for the arguments instead, but that probably can also be useful for SEA without snapshot, so I'll leave this as a TODO.

RaisinTen commented on src/node.cc:
```diff
@@ -1290,7 +1319,15 @@ static ExitCode StartInternal(int argc, char** argv) {
+                                 per_process::v8_platform.Platform(),
+                                 result->args(),
+                                 result->exec_args());
+  return main_instance.Run();
```
In the SEA config, when `useSnapshot` is set to `true` and `disableExperimentalSEAWarning` is `false`, the https://github.com/nodejs/node/blob/b5e16adb1d155759e7db405eead5a43cd425785d/lib/internal/main/embedding.js#L15 experimental SEA warning doesn't get emitted. Would it be possible to fix that?

joyeecheung replied:
Yeah I think we can fix it by adding it to the mksnapshot.js

RaisinTen commented on src/node_sea.cc:
```diff
@@ -75,9 +77,12 @@ size_t SeaSerializer::Write(const SeaResource& sea) {
   DCHECK_EQ(written_total, SeaResource::kHeaderSize);
 
   Debug("Write SEA resource code %p, size=%zu\n",
```
nit: When `useSnapshot` is enabled, this would be writing the snapshot instead of the code. What do you think about printing `snapshot` instead of `resource code` in that case?

RaisinTen commented on src/node_sea.cc:
```diff
@@ -103,7 +108,10 @@ SeaResource SeaDeserializer::Read() {
+      ReadStringView(static_cast<bool>(flags & SeaFlags::kUseSnapshot)
+                         ? StringLogMode::kAddressOnly
+                         : StringLogMode::kAddressAndContent);
   Debug("Read SEA resource code %p, size=%zu\n", code.data(), code.size());
```
nit: Same for this one, `resource code` -> `snapshot` when `useSnapshot` is `true`?

joyeecheung commented on lib/internal/main/mksnapshot.js:
```diff
@@ -167,6 +172,10 @@ function main() {
 
   const serializeMainArgs = [process, requireForUserSnapshot, minimalRunCjs];
 
+  if (isExperimentalSeaWarningNeeded()) {
```
I added a branch to print it while the SEA blob is being built - I actually think it makes more sense to not print it when the SEA is run now that we have a build phase, as this being experimental is probably only what the builder should know, but not the users (the users might not even know what Node.js is, thinking about what SEA is for, and a warning could be quite confusing for people who don't even know what Node.js is). I'll keep the warning for normal SEAs since others are already testing it, but leave it for snapshots.

RaisinTen reviewed (approved):
LGTM 🎉

piranna commented on doc/api/single-executable-applications.md:
```diff
@@ -169,14 +173,46 @@ The configuration currently reads the following top-level fields:
+### Startup snapshot support
+
+The `useSnapshot` field can be used to enable startup snapshot support. In this
+case the `main` script would not be when the final executable is launched.
```
It's missing a `run`: "would not be run"
